### ISSUE TITLE
"Action on display" code in Assessment 45632 causing "Skip logic error" alerts

### ISSUE THREAD

**Issue by rjcorwin** [Tangerine-Community/Tangerine-client#206]
Report from testing on http://databases.tangerinecentral.org/group-rti_tanzania_2016_test/_design/ojai/index.html

On Assessment 45632 (04. Classroom Observation (Kiswahili) (2016)), when you reach subtest f59efee5-67f8-c2b0-e43a-227bb1a435f4, you will see dozens of Alerts stating "Skip logic error in question xxxx ..... ReferenceError .... ResultOfQuestion could not find variable xxxx"

Asessment 45632 Backed up -> [45632-classroom-observation-kiswahili.txt](https://github.com/Tangerine-Community/Tangerine-client/files/83068/45632-classroom-observation-kiswahili.txt)
- [x] Get the big assessment into packs folder (rj)
- [x] fix the failing test on the PR #215 (chris)
- [x] write a test for this big assessment (chris)

**Comment by rjcorwin** [comment]
I’ve verified on Tanzania 2016 Test that removing the following code from “Action on display” does prevent those prior skip logic errors from popping up.

```
Utils.data("ended", true) if Utils.data("index")? and ResultOfPrevious("ended_reading_#{Utils.data('index')}") is "1"

@skip() if ResultOfPrevious("observation_type") isnt "reading" or Utils.data("ended")

# set up configuration on first run
unless Utils.data("config")?
  Utils.data("config", {
    delay : 180 * 1000 # seconds X milliseconds
  })

# try to get a previous index
index = Utils.data("index")
index = 0 unless index?

# try to get start time
start = Utils.data("start")
unless start?
  start = (new Date()).getTime()
  Utils.data("start", start)

# increase index by 1 and save
index = index + 1
Utils.data("index", index)

# hide next button
$(".next.navigation").hide()

tick = =>
  console.log "tick"
  lastStart = Utils.data('start') + ( Utils.data('config').delay * (Utils.data('index')-1) )
  elapsed = ( new Date() ).getTime() - lastStart

  # update timer
  time = moment(Utils.data('config').delay - elapsed).format("m:ss")
  if elapsed < Utils.data('config').delay
    _(vm.currentView.subtestViews[vm.currentView.index].prototypeView.questionViews).last().setHint "If the lesson has not ended, select 'The lesson continues' and wait. Your next observation will begin automatically in: #{time}"

  # check to see if we should be going next
  if elapsed >= Utils.data('config').delay
    Utils.midAlert("Next observation")
    $(".next.navigation").show()
    @next()


# automatically go to next screen
interval = setInterval( =>
  console.log "ticking"
  tick()
, 1000 )

Utils.data("interval", interval)

@onClose = => clearInterval( Utils.data("interval") )

#
# Change variable names of questions to match index
#

index = Utils.data("index")

for question in @prototypeView.questionViews
  currentName = question.getName()
  question.setName("#{currentName}_#{index}")
```

**Comment by chrisekelley** [comment]
Getting the error in console.

```
Uncaught TypeError: _this.next is not a function
```

Probably expecting it to have been implemented in the prototype. Code might work if it were:

``` javascript
@parent.next?()
```

**Comment by chrisekelley** [comment]
I implemented a next() method in SurveyRunItemView that points to the parent next:

``` javascript
  next: () ->
    console.log("next")
    @model.parent.next?()
```

Now I see an error in the displayCode:

```
vm.currentView.subtestViews
```

That is for the old view manager (vm) - we no longer use that; therefore, it is not initted. I may be able to make a stub vm and point it to 

```
@questionViews
```

Not 100% sure that would be the correct one.

Of course, s/vm.currentView.subtestViews/@questionViews in the displayCode would be better, but it is cumbersome to edit the json. Getting the edit functionality working for Tangerine-client would be super nice!

**Comment by chrisekelley** [comment]
In the js code for displayCode, we really need to make the following substitution:
    s/vm.currentView.subtestViews[vm.currentView.index].prototypeView/Tangerine.progress.currentSubview

But other subtests may use this code; it may make sense to create stubs for these to insure backward compatibility. Pondering this a bit more...

**Comment by chrisekelley** [comment]
So I believe it is working. Here is my current solution:

``` javascript
  SurveyRunItemView.prototype.onShow = function() {
    var displayCode, displaycodeFixed, error, error1, message, name, ref;
    displayCode = this.model.getString("displayCode");
    if (!_.isEmptyString(displayCode)) {
      displaycodeFixed = displayCode.replace("vm.currentView.subtestViews[vm.currentView.index].prototypeView", "Tangerine.progress.currentSubview");
      try {
        CoffeeScript["eval"].apply(this, [displaycodeFixed]);
      } catch (error1) {
        error = error1;
        name = (/function (.{1,})\(/.exec(error.constructor.toString())[1]);
        message = error.message;
        alert(name + "\n\n" + message);
        console.log("displaycodeFixed Error: " + JSON.stringify(error));
      }
    }
    return (ref = this.prototypeView) != null ? typeof ref.updateExecuteReady === "function" ? ref.updateExecuteReady(true) : void 0 : void 0;
  };
```

I am performing the substitution after fetching the string value for displayCode. I feel a tad dirty for doing this, but it would preserve compatibility with older versions. 

We should discuss if this is a terrible idea.

**Comment by rjcorwin** [comment]
That Assessment Pack is megabytes of data and also apparently not valid JSON. Darn. We need to make sure to get that thing preserved.  The earlier Tanzania data in Couch maybe...

**Comment by rjcorwin** [comment]
We have two versions of this Assessment in the packs folder now per this commit https://github.com/Tangerine-Community/Tangerine-client/commit/947c817629f355d1c3459eee6ae4b02c3bac721c

The one from 2013 and the one in 2016 Test. They seem to be different given their file size and the different IDs.

**Comment by chrisekelley** [comment]
Still working on test.

**Comment by chrisekelley** [comment]
Closed in 0.1.4.

**Comment by chrisekelley** [comment]
Test is in PR #235
